### PR TITLE
Postgresql support [BA-4759]

### PR DESCRIPTION
**Pull request by natechols**
As discussed in #4759, including migration changes and workarounds for auto-incremented columns and large objects.  I started modifying the travis-ci config but I'm probably going to need an assist there.

Note: I have been testing this with Postgresql 9.6.1, because that's what we're using here.  Other versions are TODO, it's looking like I'll have time to check v10 at least this week.

### CODE REVIEW

**Comment by natechols** [comment]
Apparently one of my migration changes doesn't work with every version of MySQL (I'm using 8.0.13, I couldn't tell what Travis-CI is using).  I think this is a no-op for Postgres anyway since there won't be any existing data to migrate.

**Comment by aednichols** [comment]
This is an amazing contribution!

I wonder whether it would be possible to make a single “create all the tables at once, in their latest configuration” migration for Postgres, given that no Cromwell instantiated prior to Postgres support could possibly have a database that needs migrating.

**Comment by natechols** [comment]
@aednichols This is certainly possible and I considered it, however @kshakir expressed a preference for sharing migrations between databases (see linked issue).

**Comment by natechols** [comment]
I have no idea why the "codecov/patch" test is failing, or how to fix it - the "details" link isn't very helpful.

**Comment by ruchim** [comment]
@natechols -- Thanks again for this amazing contribution! We're at a point of trying to wrap up reviews for some urgent changes. Our plan is to come back to this PR in 2 weeks to do a more thorough review. Sorry about the delay, but hoping to get back to this soon.

**Comment by natechols** [comment]
@ruchim That's fine, this isn't blocking me right now, as long as I have something to tell my boss.  However if someone has time to suggest what tests I should add for Postgresql, I can try to have those added by the time you're ready to review the entire mess.  (Or should it just be everything you're testing for MySQL?  This seems easy enough to add but I'm nervous about bloating your travis-ci runtimes.)

Just out of curiosity, what is the timeline for Cromwell releases this year?  I am fine using my fork for now but eventually we want to be able to use the official jarfile.

**Comment by gemmalam** [comment]
@natechols we are currently releasing every 3 weeks and next week is when we are aiming to release 41. We'll make sure to circle back to this PR and work with you on what to test when we have more time. The goal would be to get this released in 42 if all goes well!

**Comment by natechols** [comment]
@gemmalam Perfect, thank you.

**Comment by aednichols** [comment]
I happened to generate a copy of our full DDL while working on something else, I'm curious if it works on Postgres

[cromwell.sql.txt](https://github.com/broadinstitute/cromwell/files/3139919/cromwell.sql.txt)

**Comment by natechols** [comment]
@aednichols It won't, for several reasons:
1. Postgres doesn't support blob types, it uses the large object extension instead
2. There are differences in how auto-increment works - Postgres should use the serial/bigserial types
3. The identifiers will end up as lowercase if they're not quoted
It would be easy enough for me to generate the Postgres equivalent of this, however.

**Comment by natechols** [comment]
FWIW I personally find the monolithic raw CREATE statements cleaner than some of these changes to existing migrations, and it would be easy enough for me to make the switch.  But since I've already done the migration work, I don't have strong feelings about this.

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala [hunk 1]
```diff
@@ -35,7 +39,7 @@ class MetadataSlickDatabase(originalDatabaseConfig: Config)
   override def addMetadataEntries(metadataEntries: Iterable[MetadataEntry])
                                  (implicit ec: ExecutionContext): Future[Unit] = {
     val action = DBIO.seq(metadataEntries.grouped(insertBatchSize).map(dataAccess.metadataEntries ++= _).toSeq:_*)
-    runAction(action)
+    runTransaction(action)
```
Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: Why this change?

File: database/sql/src/main/scala/cromwell/database/slick/WorkflowStoreSlickDatabase.scala [hunk 2]
```diff
@@ -117,7 +121,7 @@ trait WorkflowStoreSlickDatabase extends WorkflowStoreSqlDatabase {
       })
     } yield counts.sum
     // Auto-commit mode, so each statement is individually committed to avoid deadlocks
-    runAction(action)
+    runTransaction(action)
```
Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/WorkflowStoreSlickDatabase.scala]: per the comment the non-transactionality is kinda important 🙂

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala (hunk 1 above)
Comment (natechols) [reply database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: Let me go back and double-check this.  It might be detritus from an earlier attempt that got much messier.

File: database/sql/src/main/scala/cromwell/database/slick/WorkflowStoreSlickDatabase.scala (hunk 2 above)
Comment (natechols) [reply database/sql/src/main/scala/cromwell/database/slick/WorkflowStoreSlickDatabase.scala]: This is a hard requirement of Postgres due to `importsZip: Option[Blob]` in `WorkflowStoreEntry`.  I can change it to only run transactionally for Postgres if you'd prefer, but it definitely won't work any other way.

Comment (natechols) [reply database/sql/src/main/scala/cromwell/database/slick/WorkflowStoreSlickDatabase.scala]: Actually I'd better double-check this as well.  I may have gotten carried away converting calls to run transactionally - although there are definitely at least one or two cases where it's essential.

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala (hunk 1 above)
Comment (natechols) [reply database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: Nope, this is what happens if I don't run transactionally:
```
2019-05-15 11:03:02,775 cromwell-system-akka.dispatchers.service-dispatcher-11 E
RROR - WriteMetadataActor Failed to properly process data
org.postgresql.util.PSQLException: Large Objects may not be used in auto-commit
mode.
        at org.postgresql.largeobject.LargeObjectManager.createLO(LargeObjectMan
ager.java:295)
        at org.postgresql.largeobject.LargeObjectManager.createLO(LargeObjectMan
ager.java:283)
        at org.postgresql.jdbc.PgPreparedStatement.setClob(PgPreparedStatement.j
ava:1188)
        at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.setClob(HikariPro
xyPreparedStatement.java)
        at slick.jdbc.JdbcTypesComponent$JdbcTypes$ClobJdbcType.setValue(JdbcTyp
esComponent.scala:149)
        at slick.jdbc.JdbcTypesComponent$JdbcTypes$ClobJdbcType.setValue(JdbcTyp
esComponent.scala:147)
        at slick.jdbc.OptionResultConverter.set(JdbcResultConverter.scala:33)
        at slick.jdbc.OptionResultConverter.set(JdbcResultConverter.scala:23)
        at slick.relational.ProductResultConverter.set(ResultConverter.scala:68)
        at slick.relational.ProductResultConverter.set(ResultConverter.scala:43)
        at slick.relational.TypeMappingResultConverter.set(ResultConverter.scala
:135)
        at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$MultiInsertAc
tion.$anonfun$run$15(JdbcActionComponent.scala:529)
        at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$MultiInsertAc
tion.$anonfun$run$15$adapted(JdbcActionComponent.scala:528)
        at scala.collection.Iterator.foreach(Iterator.scala:944)
        at scala.collection.Iterator.foreach$(Iterator.scala:944)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1432)
        at scala.collection.IterableLike.foreach(IterableLike.scala:71)
        at scala.collection.IterableLike.foreach$(IterableLike.scala:70)
        at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
        at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$MultiInsertAction.$anonfun$run$14(JdbcActionComponent.scala:528)
        at slick.jdbc.JdbcBackend$SessionDef.withPreparedStatement(JdbcBackend.scala:386)
        at slick.jdbc.JdbcBackend$SessionDef.withPreparedStatement$(JdbcBackend.scala:381)
        at slick.jdbc.JdbcBackend$BaseSession.withPreparedStatement(JdbcBackend.scala:448)
        at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl.preparedInsert(JdbcActionComponent.scala:501)
        at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$MultiInsertAction.run(JdbcActionComponent.scala:526)
        at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:30)
        at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:27)
        at slick.dbio.DBIOAction$$anon$4.$anonfun$run$3(DBIOAction.scala:239)
        at scala.collection.Iterator.foreach(Iterator.scala:944)
        at scala.collection.Iterator.foreach$(Iterator.scala:944)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1432)
        at scala.collection.IterableLike.foreach(IterableLike.scala:71)
        at scala.collection.IterableLike.foreach$(IterableLike.scala:70)
        at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
        at slick.dbio.DBIOAction$$anon$4.run(DBIOAction.scala:239)
        at slick.dbio.DBIOAction$$anon$4.run(DBIOAction.scala:237)
        at slick.dbio.SynchronousDatabaseAction$$anon$7.run(DBIOAction.scala:486)
        at slick.basic.BasicBackend$DatabaseDef$$anon$2.liftedTree1$1(BasicBackend.scala:275)
        at slick.basic.BasicBackend$DatabaseDef$$anon$2.run(BasicBackend.scala:275)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```
I'm convinced there is a bug in how `Clob` is handled by the Postgres driver, because it's just a `text` field in tables, which shouldn't require large object support.  Unfortunately I don't think there's an elegant workaround the way there was for the `SELECT` queries.

File: database/sql/src/main/scala/cromwell/database/slick/WorkflowStoreSlickDatabase.scala (hunk 2 above)
Comment (natechols) [reply database/sql/src/main/scala/cromwell/database/slick/WorkflowStoreSlickDatabase.scala]: This one does appear to be unnecessary, so I pushed a revert.

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala (hunk 1 above)
Comment (natechols) [reply database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: Should I change it to run transactionally only for Postgres?

Comment (aednichols) [reply database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: Yes, please.

Transactions have semantic and performance effects. If the semantics don't require a transaction, I'd steer clear. We insert a ton of metadata and making the RDBMS do so in a transaction basically tells it, stage all of these things in a temporary area such that they can be committed with an atomic switch - instead of processing them as they come in. One can imagine that the latter is a lot easier.

Comment (natechols) [reply database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: Okay, done.

**Comment by natechols** [comment]
@gemmalam Just wanted to check up on this since it's been a month - will it be reviewed as part of the current sprint, and hopefully included in release 43?

**Comment by gemmalam** [comment]
Hi @natechols! We are working on adding testing for this support. You can check out the PR https://github.com/broadinstitute/cromwell/pull/5018

Completing the reviewing and approval of this PR did carry over from our last sprint to this one. We are currently in our first week of a three week sprint.

**Comment by natechols** [comment]
@gemmalam Sounds good, please let me know if there's anything I can do to help.

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala [hunk 3]
```diff
@@ -133,7 +138,7 @@ class MetadataSlickDatabase(originalDatabaseConfig: Config)
         updateCount <- dataAccess.
           customLabelEntriesForWorkflowExecutionUuidAndLabelKey(
             (customLabelEntry.workflowExecutionUuid, customLabelEntry.customLabelKey)
-          ).update(customLabelEntry)
+          ).update(customLabelEntry.toUpdateColumns)
```
Comment (natechols) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: @kshakir The specific error that motivated this change (and several others) was this:
```
2019-06-17 15:07:00,298 cromwell-system-akka.dispatchers.service-dispatcher-94 ERROR - Error summarizing metadata
org.postgresql.util.PSQLException: ERROR: null value in column "CUSTOM_LABEL_ENTRY_ID" violates not-null constraint
  Detail: Failing row contains (null, cromwell-workflow-id, cromwell-a8afcb27-4411-4c83-9b38-8402a23cc579, a8afcb27-4411-4c83-9b38-8402a23cc579).
        at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2440)
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2183)
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:308)
        at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:441)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:365)
        at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:143)
        at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:120)
        at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61)
        at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java)
        at slick.jdbc.JdbcActionComponent$UpdateActionExtensionMethodsImpl$$anon$8.$anonfun$run$10(JdbcActionComponent.scala:325)
        at slick.jdbc.JdbcActionComponent$UpdateActionExtensionMethodsImpl$$anon$8.$anonfun$run$10$adapted(JdbcActionComponent.scala:321)
        at slick.jdbc.JdbcBackend$SessionDef.withPreparedStatement(JdbcBackend.scala:386)
        at slick.jdbc.JdbcBackend$SessionDef.withPreparedStatement$(JdbcBackend.scala:381)
        at slick.jdbc.JdbcBackend$BaseSession.withPreparedStatement(JdbcBackend.scala:448)
        at slick.jdbc.JdbcActionComponent$UpdateActionExtensionMethodsImpl$$anon$8.run(JdbcActionComponent.scala:321)
        at slick.jdbc.JdbcActionComponent$UpdateActionExtensionMethodsImpl$$anon$8.run(JdbcActionComponent.scala:320)
        at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:30)
        at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:27)
        at slick.basic.BasicBackend$DatabaseDef$$anon$2.liftedTree1$1(BasicBackend.scala:275)
        at slick.basic.BasicBackend$DatabaseDef$$anon$2.run(BasicBackend.scala:275)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```
(This is also what prevents us from doing "upserts" in Postgres.)

Comment (natechols) [reply database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: Also:
```
cromwell=# \d "CUSTOM_LABEL_ENTRY";
                                                    Table "public.CUSTOM_LABEL_ENTRY"
         Column          |          Type          |                                      Modifiers
-------------------------+------------------------+--------------------------------------------------------------------------------------
 CUSTOM_LABEL_ENTRY_ID   | bigint                 | not null default nextval('"CUSTOM_LABEL_ENTRY_CUSTOM_LABEL_ENTRY_ID_seq"'::regclass)
 CUSTOM_LABEL_KEY        | character varying(255) | not null
 CUSTOM_LABEL_VALUE      | character varying(255) | not null
 WORKFLOW_EXECUTION_UUID | character varying(100) | not null
Indexes:
    "PK_CUSTOM_LABEL_ENTRY" PRIMARY KEY, btree ("CUSTOM_LABEL_ENTRY_ID")
    "UC_CUSTOM_LABEL_ENTRY_CLK_CLV_WEU" UNIQUE CONSTRAINT, btree ("CUSTOM_LABEL_KEY", "CUSTOM_LABEL_VALUE", "WORKFLOW_EXECUTION_UUID")
    "IX_CUSTOM_LABEL_ENTRY_CLK_CLV" btree ("CUSTOM_LABEL_KEY", "CUSTOM_LABEL_VALUE")
    "SYS_IDX_11226" btree ("WORKFLOW_EXECUTION_UUID")
Foreign-key constraints:
    "FK_CUSTOM_LABEL_ENTRY_WORKFLOW_EXECUTION_UUID" FOREIGN KEY ("WORKFLOW_EXECUTION_UUID") REFERENCES "WORKFLOW_METADATA_SUMMARY_ENTRY"("WORKFLOW_EXECUTION_UUID") ON DELETE CASCADE```

Comment (kshakir) [reply database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: Re PostgreSQL+Slick/Liquibase:

The other DSP-Batch'ers and I discussed changing the old liquibase changelogs, modifying them to be able to run on both PostgreSQL+MySQL+HSQLDB. During the discussion I was convinced to go another way than what I already told you.

I left a few comments over in this example commit: https://github.com/broadinstitute/cromwell/commit/77630aaa90cb27af9c5d9360a5e8a2c0660a1e10

(For those who come across this message in the future, the commit may disappear but the gist of the changes should make it into this PR.)

The commit goes about explicitly-including supported databases. What's preferred with this approach is:
1. No changes to the old MD5s
2. If needed we can do the same thing for additional DBMS (ex: MariaDB)

Using liquibase's changelog generator (plus some minor tweaks) I've partially recreated and attached the schemas from MySQL. Hopefully the XML files are a good starting point for PostgreSQL.

[engine_schema.xml](https://github.com/broadinstitute/cromwell/files/3308568/engine_schema.xml.txt)
[metadata_schema.xml](https://github.com/broadinstitute/cromwell/files/3308569/metadata_schema.xml.txt)

Later, towards the end of the commit, hopefully the examples in the package `cromwell.database` will workaround the issues with AutoInc and Lob columns.

Comment (natechols) [reply database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: @kshakir Thanks, I'll take a look at 77630aa.  Should I try incorporating this into my PR?

Comment (kshakir) [reply database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: Yes please, do try replicating the example changes across your PR. Thanks again for your patience with all of this.

Comment (natechols) [reply database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: I've just started on the migrations, but the Scala code changes should be complete now if you want to start reviewing them.  So far everything is working perfectly for me in my local environment.

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala [hunk 4]
```diff
@@ -34,7 +34,8 @@ class MetadataSlickDatabase(originalDatabaseConfig: Config)
   override def addMetadataEntries(metadataEntries: Iterable[MetadataEntry])
                                  (implicit ec: ExecutionContext): Future[Unit] = {
     val action = DBIO.seq(metadataEntries.grouped(insertBatchSize).map(dataAccess.metadataEntries ++= _).toSeq:_*)
-    runAction(action)
+    // workaround for large object support
+    if (isPostgresql) runTransaction(action) else runAction(action)
```
Comment (kshakir) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: Today-I-Learned.

1. Please also leave a link pointing at: https://stackoverflow.com/questions/3164072/large-objects-may-not-be-used-in-auto-commit-mode#answer-3164352
2. I'm thinking this could be consolidated into a reusable `def runLobAction()` method that does this `if` statement, a sibling of `def runAction()` and `def runTransaction()`

Comment (natechols) [reply database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: done

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala (hunk 3 above)
Comment (natechols) [reply database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: The migrations are now done too.  I find this much cleaner than what I had before.

**Comment by pb-cdunn** [comment]
@natechols, Very nice!

File: database/migration/src/main/resources/changesets/postgresql_engine_schema.xml [hunk 5]
```diff
@@ -0,0 +1,326 @@
+<?xml version="1.1" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog objectQuotingStrategy="QUOTE_ALL_OBJECTS"
+                   xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.5.xsd">
+
+    <changeSet author="kshakir (generated)" id="postgresql-engine-init-1" dbms="postgresql">
+        <createTable tableName="CALL_CACHING_AGGREGATION_ENTRY">
+            <column autoIncrement="true" name="CALL_CACHING_AGGREGATION_ENTRY_ID" type="SERIAL">
```
Comment (natechols) [inline database/migration/src/main/resources/changesets/postgresql_engine_schema.xml]: FYI, I was worried that `autoIncrement="true"` would conflict with `type="SERIAL"`, since the separate `addAutoIncrement` instructions in the old migrations had this problem, but it appears this particular combination is just fine.

**Comment by natechols** [comment]
> There is one I'm having trouble googling a fix for. I can't figure out how to shut off PostgreSQL exceptions printing possibly sensitive row contents via their messages.

I wouldn't be surprised if this is baked into the JDBC layer.  We could try something like this:
```
diff --git a/database/sql/src/main/scala/cromwell/database/slick/SlickDatabase.scala b/database/sql/src/main/scala/cromwell/database/slick/SlickDatabase.scala
index 5d28cf1..5b0e227 100644
--- a/database/sql/src/main/scala/cromwell/database/slick/SlickDatabase.scala
+++ b/database/sql/src/main/scala/cromwell/database/slick/SlickDatabase.scala
@@ -11,6 +11,7 @@ import net.ceedubs.ficus.Ficus._
 import org.slf4j.LoggerFactory
 import slick.basic.DatabaseConfig
 import slick.jdbc.{JdbcCapabilities, JdbcProfile, TransactionIsolation}
+import org.postgresql.util.{PSQLException, ServerErrorMessage}

 import scala.concurrent.duration._
 import scala.concurrent.{Await, ExecutionContext, Future}
@@ -199,6 +200,8 @@ abstract class SlickDatabase(override val originalDatabaseConfig: Config) extend
             case _ => /* keep going */
           }
           throw rollbackException
+        case pe: PSQLException =>
+          throw new PSQLException(new ServerErrorMessage(s"Oh no, a postgres error occurred!  ${pe.getMessage}"))
       }
     }(actionExecutionContext)
   }
```
only with some on-the-fly modification of the error message instead of my dummy string.  This compiles for me, but I'm not sure how to test it.

**Comment by kshakir** [comment]
Ah, oh well, thanks for the second set of eyes.

https://github.com/broadinstitute/cromwell/commit/2682c001d99823098e655acd1dd7a3062a68f495 has your change implemented with some ~nasty~ reflection that hopefully the rest of the DSP-Batcher's will accept, and the test re-enabled.

CI running here to see if it breaks anything:
https://travis-ci.com/broadinstitute/cromwell/builds/116462548

Assuming this works, I think the changes will pass all of our existing unit and integration tests!

We can get two others to review. I should abstain due to our collaboration on the code. 😉

**Comment by natechols** [comment]
Should I merge 2682c00 with this if the CI tests pass?

**Comment by kshakir** [comment]
> Should I merge 2682c00 with this if the CI tests pass?

Yes plz.

**Comment by kshakir** [comment]
Tests continue to run well. I forgot about documentation though, both a short blurb in the `/CHANGELOG.md`, and that linking over to as much documentation as needed under `/docs`.

Not sure if you've used MkDocs before, but [the live docs are hosted here](https://cromwell.readthedocs.io/en/stable/search.html?q=mysql), and the docs can be tested locally by running `mkdocs serve` from the cromwell root directory and then browsing to http://localhost:8000

If you run into any issues drop us a message here.

Thanks!

File: CHANGELOG.md [hunk 6]
```diff
@@ -55,6 +55,13 @@ YAML](https://cromwell.readthedocs.io/en/stable/Configuring/#yaml) for more info
 * It is now possible to use `includeKey` and `excludeKey` at the same time. If so, the metadata key must match the `includeKey` **and not** match the `excludeKey` to be included.
 * It is now possible to use "`calls`" as one of your `excludeKey`s, to request that only workflow metadata gets returned.
 
+### Postgresql support
+
+Cromwell now supports Postgresql (version 9.6 or higher, with the Large Object
```
Comment (aednichols) [inline CHANGELOG.md]:
```suggestion
Cromwell now supports PostgreSQL (version 9.6 or higher, with the Large Object
```
Sorry

**Comment by kshakir** [comment]
@natechols Do you mind making the last change requested above, then git-squashing everything into a one commit with a single commit message? Thanks!

**Comment by natechols** [comment]
@kshakir done, I think, but someone should probably double check as I am not 100% confident in my git abilities.

**Comment by natechols** [comment]
I don't understand the Travis CI failure - is this unrelated?  Is there a way for me to re-run the test?

**Comment by aednichols** [comment]
It's a dependency resolution error, which seems surprising. I cleared Travis cache and restarted.

**Comment by aednichols** [comment]
I'm hearing reports within our org that there is a systemic problem with pulling artifacts
